File: Functions/plotAgeingPanel.py
"""
plotAgeingPanel - Draw one ageing-overview panel (capacity or resistance)
into a matplotlib axes.

Python counterpart of Functions/plotAgeingPanel.m, used by
python_scripts/PlotAgeingCombinedOverview.py (todo #101).

Plots every cell in cell_list_all as a faint grey background trace, then the
highlighted cells (cell_plot_list) as coloured traces, so multi-panel
figures can compose several such calls with a shared legend.

Author: GitHub Copilot (for Robinson Medina / Feye Hoekstra)
Date: 2026-08-24

Compliance: R-004, R-017.
"""

import numpy as np

_LW_DATA = 1.0                  # R-017: data traces 1.0 pt
_GREY = (0.75, 0.75, 0.75)      # faint background cells


def _local_xy(cur, x_mode, y_mode, fec_col, resistance_scale):
    """Compute (x, y) arrays for one cell's rows, per x_mode/y_mode."""

    if x_mode == 'time':
        x = (cur['Timestamp'] - cur['Timestamp'].iloc[0]).dt.total_seconds().values / 86400.0
    elif x_mode == 'fec':
        x = cur[fec_col].values
    else:
        raise ValueError("x_mode must be 'time' or 'fec'.")

    if y_mode == 'capacity':
        y = cur['Capacity [Ah]'].values / cur['Capacity [Ah]'].values[0]
    elif y_mode == 'resistance':
        y = cur['Resistance [Ohm]'].values * resistance_scale
    else:
        raise ValueError("y_mode must be 'capacity' or 'resistance'.")

    return x, y


def plot_ageing_panel(ax, df, cell_list_all, cell_plot_list, color_list, marker_list,
                       x_mode, y_mode, resistance_scale=1):
    """
    Draw one ageing-overview panel (capacity or resistance) into ax.

    Parameters
    ----------
    ax : matplotlib.axes.Axes
    df : pandas.DataFrame
        Preprocessed table with columns 'cell_number', 'Timestamp',
        'Capacity [Ah]'/'Resistance [Ohm]', and the FEC column.
    cell_list_all : list of str
        All cell names to draw faint in the background.
    cell_plot_list : list of str
        Highlighted cell names (draw order == legend order).
    color_list : list of RGB tuples, one per highlighted cell.
    marker_list : list of matplotlib line styles, one per highlighted cell.
    x_mode : 'time' or 'fec'.
    y_mode : 'capacity' or 'resistance'.
    resistance_scale : float
        Multiplier for resistance (e.g. 1000 for mOhm); ignored for capacity.

    Returns
    -------
    handles : list of matplotlib.lines.Line2D
        Highlighted line handles (for a shared legend).
    max_x_highlight : float
        Max x across highlighted cells (for a common xlim).
    """

    if y_mode == 'capacity':
        fec_col = 'CheckupCapacityFEC'
    elif y_mode == 'resistance':
        fec_col = 'CheckupResistanceFEC'
    else:
        raise ValueError('y_mode must be capacity or resistance.')

    # Faint background: every cell in the dataset.
    for cell_name in cell_list_all:
        cur = df[df['cell_number'] == cell_name]
        if cur.empty:
            continue
        x, y = _local_xy(cur, x_mode, y_mode, fec_col, resistance_scale)
        ax.plot(x, y, '-', linewidth=_LW_DATA, color=_GREY, zorder=1)

    # Highlighted cells (deterministic colour/order for the shared legend).
    handles = []
    max_x_highlight = 0.0
    for cell_name, color, marker in zip(cell_plot_list, color_list, marker_list):
        cur = df[df['cell_number'] == cell_name]
        if cur.empty:
            print(f'plot_ageing_panel: Cell {cell_name} not found.')
            continue
        x, y = _local_xy(cur, x_mode, y_mode, fec_col, resistance_scale)
        (h,) = ax.plot(x, y, marker, linewidth=_LW_DATA, color=color, zorder=2)
        handles.append(h)
        if len(x) > 0:
            max_x_highlight = max(max_x_highlight, np.max(x))

    return handles, max_x_highlight


__all__ = ['plot_ageing_panel']

File: Functions/plotAgeingPanel.m
function [handles, maxXHighlight] = plotAgeingPanel(ax, df, cellListAll, cellPlotList, colorList, markerList, xMode, yMode, resistanceScale)
%PLOTAGEINGPANEL Draw one ageing-overview panel (capacity or resistance) into an axis.
%   Plots every cell in cellListAll as a faint grey background trace, then the
%   highlighted cells (cellPlotList) as coloured traces. Shared by the merged
%   capacity-degradation / resistance-increase ageing figures.
%
%   Inputs:
%     ax             target axes (already created/positioned by the caller)
%     df             preprocessed table (cell_number, Timestamp, quantity + FEC cols)
%     cellListAll    all cell names to draw faint in the background
%     cellPlotList   highlighted cell names (draw order == legend order)
%     colorList      cell array of RGB triplets, one per highlighted cell
%     markerList     cell array of line styles, one per highlighted cell
%     xMode          'time' -> days since first sample; 'fec' -> equivalent cycles
%     yMode          'capacity' -> C/C0 (dimensionless); 'resistance' -> Ohm*scale
%     resistanceScale multiplier for resistance (e.g. 1000 for mOhm); default 1
%
%   Outputs:
%     handles        highlighted line handles (for a shared legend)
%     maxXHighlight  max x across highlighted cells (for a common xlim)
%
% Author: NEXTBMS Team
% Date: 2026-08-12

    if nargin < 9 || isempty(resistanceScale); resistanceScale = 1; end

    switch lower(yMode)
        case 'capacity';   fecCol = 'CheckupCapacityFEC';
        case 'resistance'; fecCol = 'CheckupResistanceFEC';
        otherwise; error('plotAgeingPanel:yMode', 'yMode must be capacity or resistance.');
    end

    LW_DATA = 1.0;                 % R-017: data traces 1.0 pt
    greyCol = [0.75 0.75 0.75];    % faint background cells

    hold(ax, 'on');

    % Faint background: every cell in the dataset.
    for i = 1:numel(cellListAll)
        cur = df(strcmp(df.('cell_number'), cellListAll{i}), :);
        if isempty(cur); continue; end
        [x, y] = localXY(cur, xMode, yMode, fecCol, resistanceScale);
        plot(ax, x, y, '-', 'LineWidth', LW_DATA, 'Color', greyCol, 'HandleVisibility', 'off');
    end

    % Highlighted cells (deterministic colour/order for the shared legend).
    handles = gobjects(1, numel(cellPlotList));
    maxXHighlight = 0;
    for idx = 1:numel(cellPlotList)
        cur = df(strcmp(df.('cell_number'), cellPlotList{idx}), :);
        if isempty(cur)
            warning('plotAgeingPanel:missingCell', 'Cell %s not found.', cellPlotList{idx});
            continue;
        end
        [x, y] = localXY(cur, xMode, yMode, fecCol, resistanceScale);
        handles(idx) = plot(ax, x, y, markerList{idx}, 'LineWidth', LW_DATA, 'Color', colorList{idx});
        if ~isempty(x); maxXHighlight = max(maxXHighlight, max(x)); end
    end
    handles = handles(isgraphics(handles));
end

function [x, y] = localXY(cur, xMode, yMode, fecCol, resistanceScale)
    switch lower(xMode)
        case 'time'; x = days(cur.('Timestamp') - cur.('Timestamp')(1));
        case 'fec';  x = cur.(fecCol);
        otherwise; error('plotAgeingPanel:xMode', 'xMode must be time or fec.');
    end
    switch lower(yMode)
        case 'capacity';   y = cur.('Capacity [Ah]') / cur.('Capacity [Ah]')(1);
        case 'resistance'; y = cur.('Resistance [Ohm]') * resistanceScale;
    end
end
